File: tests/example02_hello_world.py
'''test functionality in Python'''
import maf

from mpi4py.MPI import COMM_WORLD # this lets mpi4py handle MPI_Init and MPI_Finalize

import sys
import cPickle
import timeit

class HelloWorldAction(maf.Action):

    def __init__(self, my_string='Hello World!', my_int=0, my_float=0.0):
        maf.Action.__init__(self)
        self.my_string = my_string
        self.my_int = my_int
        self.my_float = my_float

    def run(self):
        maf.log('{} my_int: {} my_float: {}'.format(self.my_string, self.my_int, self.my_float))
    
    def serialize(self, archive):
        self.my_string = archive.rw_string(self.my_string)
        self.my_int = archive.rw_long_long(self.my_int)
        self.my_float = archive.rw_double(self.my_float)


class PickleSerialer(HelloWorldAction):

    def serialize(self, archive):
        data = dict(self.__dict__)
        if 'this' in data:
            del data['this'] # don't pickle the SWIG object
        string = cPickle.dumps(data)
        string = archive.rw_string(string)
        self.__dict__.update(cPickle.loads(string))


class TestBcastController(maf.BcastController):
    
    def __init__(self, action_class):
        maf.BcastController.__init__(self)
        self.action_class = action_class

    def main(self):
        act = self.action_class()
        self.distribute(act).run()
        act2 = self.action_class('Whatever', 2, 44)
        self.distribute(act2).run()
        act = self.action_class('The final act!', sys.maxint, sys.float_info.max)
        self.distribute(act).run()


maf.register(HelloWorldAction)
maf.register(PickleSerialer)

start = timeit.default_timer()
controller = TestBcastController(HelloWorldAction)
controller.start()
serializer_time = timeit.default_timer() - start

COMM_WORLD.Barrier()

start = timeit.default_timer()
controller = TestBcastController(PickleSerialer)
controller.start()
pickle_time = timeit.default_timer() - start

maf.log('SERIALIZE time: {}\n'
    '    PICKLE time   : {}'.format(serializer_time, pickle_time))

File: tests/Example02HelloWorld.cpp
#include "maf/maf.hpp"

#include <mpi.h>

class HelloWorldAction : public maf::Action {

public:

    std::string my_string;
    int my_int;
    float my_float;
    double my_double;

    HelloWorldAction() : maf::Action() {

    };

    HelloWorldAction(std::string string_, int int_, float float_, double double_) : maf::Action() {
        this->my_string = string_;
        this->my_int = int_;
        this->my_float = float_;
        this->my_double = double_;
    };

    void run() {
        maf::mpi_print("Hello World! (from c++)",
            "\n     my_string: `", this->my_string,
            "`\n     my_int: `", this->my_int,
            "`\n     my_float: `", this->my_float,
            "`\n     my_double: `", this->my_double, "`"
            );
    };

    void serialize(std::shared_ptr<maf::Archive> archive) {
        (*archive) & this->my_string;
        (*archive) & this->my_int;
        (*archive) & this->my_float;
        (*archive) & this->my_double;
    };

    std::string type_name() {
        return "HelloWorldAction";
    }

};

class TestBcastController : public maf::BcastController {
public:
    void main() {
        std::shared_ptr<maf::Action> act = std::shared_ptr<maf::Action>(new HelloWorldAction);
        this->distribute(act);
        act->run();
        act = std::shared_ptr<maf::Action>(new HelloWorldAction("string 1", 2, 3.3, 4.4));
        this->distribute(act);
        act->run();
        act = std::shared_ptr<maf::Action>(new HelloWorldAction("string 2", 4, 66.66, 88.88e88));
        this->distribute(act);
        act->run();
    };
};

int main(int argc, char* argv[]) {
    MPI_Init(&argc, &argv);

    try {
        // std::shared_ptr<ActionFactory> factory((ActionFactory*)(new TActionFactory<HelloWorldAction>("HelloWorldAction")));
        auto factory = std::shared_ptr<maf::ActionFactory>((maf::ActionFactory*)(new maf::TActionFactory<HelloWorldAction>("HelloWorldAction")));
        maf::ActionFactory::Register(factory);

        TestBcastController controller;
        controller.start();
    }
    catch (std::exception* ex) {
        maf::mpi_print("FAILED: ", ex->what());
    }
    catch (...) {
        maf::mpi_print("FAILED: no idea what happened");
    }

    MPI_Finalize();
}
